tasks/ReedSwitchSimulatorTask.cpp

```diff
@@ -0,0 +1,128 @@
+
+#include "ReedSwitchSimulatorTask.hpp"
+
+using namespace std;
+using namespace base;
+using namespace gazebo_usv;
+using namespace linux_gpios;
+
+ReedSwitchSimulatorTask::ReedSwitchSimulatorTask(std::string const& name)
+    : ReedSwitchSimulatorTaskBase(name)
+{
+}
+
+ReedSwitchSimulatorTask::~ReedSwitchSimulatorTask()
+{
+}
+
+/// The following lines are template definitions for the various state machine
+// hooks defined by Orocos::RTT. See ReedSwitchSimulatorTask.hpp for more detailed
+// documentation about them.
+
+bool ReedSwitchSimulatorTask::configureHook()
+{
+    if (!ReedSwitchSimulatorTaskBase::configureHook()) {
+        return false;
+    }
+    m_revolution_time = _revolution_time.get();
+    m_trigger_time = _trigger_time.get();
+    return true;
+}
+
+bool ReedSwitchSimulatorTask::startHook()
+{
+    if (!ReedSwitchSimulatorTaskBase::startHook()) {
+        return false;
+    }
+    return true;
+}
+
+void ReedSwitchSimulatorTask::updateTimeInCycle(int64_t delta)
+{
+    m_timer += delta;
+    if (m_timer < 0){
+        m_timer += m_revolution_time.microseconds;
+    }
+    m_timer = m_timer % m_revolution_time.microseconds;
+}
+
+CommandDirection motorCommand(GPIOState const& extend, GPIOState const& retract)
+{
+    bool extend_state = extend.states[0].data;
+    bool retract_state = retract.states[0].data;
+    if (!extend_state && !retract_state) {
+        return CommandDirection{0, Time::now()};
+    }
+    else if (extend_state && !retract_state) {
+        return CommandDirection{1, Time::now()};
+    }
+    else if (!extend_state && retract_state) {
+        return CommandDirection{-1, Time::now()};
+    }
+    else {
+        throw std::runtime_error("Invalid command direction - Both gpio states should "
+                                 "not be true at the same time.");
+    }
+}
+
+void ReedSwitchSimulatorTask::updateMotorCommandDirection(GPIOState const& extend,
+    GPIOState const& retract)
+{
+    CommandDirection command_direction = motorCommand(extend, retract);
+    if (m_last_command_direction.has_value()) {
+        Time delta_time = Time::now() - m_last_command_direction->time;
+        updateTimeInCycle(delta_time.microseconds * m_last_command_direction->direction);
+    }
+    m_last_command_direction = command_direction;
+}
+
+GPIOState GPIO(bool flag)
+{
+    auto now = Time::now();
+    GPIOState gpio;
+    gpio.time = now;
+    gpio.states.resize(1);
+    gpio.states[0].time = now;
+    gpio.states[0].data = flag;
+    return gpio;
+}
+
+bool ReedSwitchSimulatorTask::sensorActive()
+{
+    return m_timer <= m_trigger_time.microseconds;
+}
+
+void ReedSwitchSimulatorTask::updateHook()
+{
+    ReedSwitchSimulatorTaskBase::updateHook();
+
+    GPIOState gpio_retract;
+    if (_gpio_retract.read(gpio_retract) == RTT::NoData) {
+        return;
+    }
+    GPIOState gpio_extend;
+    if (_gpio_extend.read(gpio_extend) == RTT::NoData) {
+        return;
+    }
```

Comment:
validate taht the commands have the correct size

```suggestion
    GPIOState gpio_retract;
    if (_gpio_retract.read(gpio_retract) == RTT::NoData) {
        return;
    }
    GPIOState gpio_extend;
    if (_gpio_extend.read(gpio_extend) == RTT::NoData) {
        return;
    }
```